Bitbucket.Authentication/OAuth/OAuthAuthenticator.cs

```diff
@@ -0,0 +1,321 @@
+            var authCode = qs.Keys.Where(k => k.EndsWith("code", StringComparison.InvariantCultureIgnoreCase)).Select(k => qs[k]).FirstOrDefault();
+
+
+            if (string.IsNullOrWhiteSpace(authCode))
+            {
+                var error_desc = qs["error_description"];
+                throw new Exception("Request for an OAuth request_token was denied" + error_desc);
+            }
+
+            return authCode;
+        }
+
+        /// <summary>
+        ///     Use a request_token to get an access_token
+        /// </summary>
+        /// <param name="targetUri"></param>
+        /// <param name="authCode"></param>
+        /// <returns></returns>
+        private async Task<AuthenticationResult> GetAccessToken(TargetUri targetUri, string authCode)
+        {
+            Token token = null;
+            Token refreshToken = null;
+
+            using (HttpClientHandler handler = targetUri.HttpClientHandler)
+            {
+                using (HttpClient httpClient = new HttpClient(handler)
+                {
+                    Timeout = TimeSpan.FromMilliseconds(RequestTimeout)
+                })
+                {
+                    var tokenUrl = string.Format(
+                       TokenUri +
+                       "?grant_type=authorization_code&code={0}&client_id={1}&client_secret={2}&state=authenticated",
+                       authCode,
+                       ConsumerKey,
+                       ConsumerSecret);
+                    var url = new Uri(new Uri(targetUri.ToString()), tokenUrl).AbsoluteUri;
+
+                    var content = new MultipartFormDataContent();
+                    content.Add(new StringContent("authorization_code"), "grant_type");
+                    content.Add(new StringContent(authCode), "code");
+                    content.Add(new StringContent(ConsumerKey), "client_id");
+                    content.Add(new StringContent(ConsumerSecret), "client_secret");
+                    content.Add(new StringContent("authenticated"), "state");
+                    content.Add(new StringContent(CallbackUri), "redirect_uri");
+
+                    using (HttpResponseMessage response = await httpClient.PostAsync(url, content))
+                    {
+                        Trace.WriteLine($"server responded with {response.StatusCode}.");
+
+                        switch (response.StatusCode)
+                        {
+                            case HttpStatusCode.OK:
+                            case HttpStatusCode.Created:
+                                {
+                                    string responseText = await response.Content.ReadAsStringAsync();
+
+                                    Match tokenMatch;
+                                    if ((tokenMatch = Regex.Match(responseText, @"\s*""access_token""\s*:\s*""([^""]+)""\s*", RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase)).Success
```

Comment:
Seems like there's a lot of dynamic regular expressions happen here. I'm guessing that this will have some performance impact. Likely negligible compared to the cost of network operations, but I wanted to point it out.